bash-snippets.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "BASH\-SNIPPETS" "1" "April 2020" "" ""
.
.SH "NAME"
\fBbash\-snippets\fR \- A collection of small bash scripts for heavy terminal users
.
.SH "Tools"
.
.SS "Cheat"
Description: Cheatsheets for quick information about multiple programming languages along with terminal commands
.
.P
Usage: cheat [flags] [command] or cheat [flags] [programming language] [subject]
.
.IP "\(bu" 4
\-s Does a search for last argument rather than looking for exact match
.
.IP "\(bu" 4
\-i Case insensitive search
.
.IP "\(bu" 4
\-b Word boundaries in search
.
.IP "\(bu" 4
\-r Recursive search
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Special Pages:
.
.IP "\(bu" 4
hello Describes building the hello world program written in the language
.
.IP "\(bu" 4
list This lists all cheatsheets related to previous arg if none it lists all cheatsheets
.
.IP "\(bu" 4
learn Shows a learn\-x\-in\-minutes language cheat sheet perfect for getting started with the language
.
.IP "\(bu" 4
1line A collection of one\-liners in this language
.
.IP "\(bu" 4
weirdness A collection of examples of weird things in this language
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
cheat rust hello
.
.IP "\(bu" 4
cheat \-r \-b \-i go
.
.IP "\(bu" 4
cheat julia Functions
.
.IP "\(bu" 4
cheat \-i go operators
.
.IP "" 0
.
.SS "Cloudup"
Description: Backs up a users github repositories to your bitbucket account\. With no flags cloudup will guide you through backing up a single repository
.
.P
Usage: cloudup [flags] or cloudup [flags] [listOfGHRepoNamesSplitBySpaces]
.
.IP "\(bu" 4
\-p Upload the repositor(y)(ies) as private to bitbucket (must have private repo ability on bitbucket)
.
.IP "\(bu" 4
\-a Backup all github repositories
.
.IP "\(bu" 4
\-s Only backup repositories that you have created (no forks) (only works in conjunction with the \-a flag)
.
.IP "\(bu" 4
\-t Backup the repository with a timestamp added to the repostiory name (will always create a new unique bitbucket repo)
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
cloudup
.
.IP "\(bu" 4
cloudup \-p \-a
.
.IP "\(bu" 4
cloudup \-p nameOfRepo1 nameOf Repo2
.
.IP "\(bu" 4
cloudup nameOfRep
.
.IP "" 0
.
.SS "Crypt"
Description: A wrapper around openssl that facilitates encrypting and decrypting files\.
.
.P
Usage: crypt [flag] [inputFile] [outputFile]
.
.IP "\(bu" 4
\-e Encrypt the inputFile and store it in the outputFil
.
.IP "\(bu" 4
\-d Decrypt the inputFile and store it in the outputFil
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
crypt \-e mySecretFile\.txt myEncryptedFile\.jpg (change filetype so default program is incorrect)
.
.IP "\(bu" 4
crypt \-d myEncryptedFile\.jpg thisIsNowDecrypted\.txt (change filetype back so now default program is correct)
.
.IP "" 0
.
.SS "CryptoCurrency"
Description: A realtime cryptocurrency converter\. With no flags it will pull down the latest stats of the top 10 cryptos
.
.P
Usage: cryptocurrency or cryptocurrency [flag] or cryptocurrency [flag] [arg]
.
.IP "\(bu" 4
\-o Utilize the old functionality of the tool
.
.IP "\(bu" 4
\-f Fiat currency for conversions
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples: cryptocurrency \-o cryptocurrency \-f eur
.
.SS "Currency"
Description: A realtime currency converter\. With no flags it will guide you through the currency exchange
.
.P
Usage: currency or currency [flag] or currency [base] [exchangeTo] [amount]
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Supported Currencies:
.
.P
| AUD | BGN | BRL | CAD | | CHF | CNY | CZK | DKK | | EUR | GBP | HKD | HRK | | HUF | IDR | ILS | INR | | JPY | KRW | MXN | MYR | | NOK | NZD | PHP | PLN | | RON | RUB | SEK | SGD | | THB | TRY | USD | ZAR |
.
.P
Examples:
.
.IP "\(bu" 4
currency EUR USD 12\.35
.
.IP "\(bu" 4
currency
.
.IP "" 0
.
.SS "Geo"
Description: Provides quick access for wan, lan, router, dns, mac, and ip geolocation data
.
.P
Usage: geo [flag]
.
.IP "\(bu" 4
\-w Returns WAN IP
.
.IP "\(bu" 4
\-l Returns LAN IP(s)
.
.IP "\(bu" 4
\-r Returns Router IP
.
.IP "\(bu" 4
\-d Returns DNS Nameserver
.
.IP "\(bu" 4
\-m Returns MAC address for interface\. Ex\. eth0
.
.IP "\(bu" 4
\-g Returns Current IP Geodata
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
geo \-g
.
.IP "\(bu" 4
geo \-wlrdgm eth0
.
.IP "" 0
.
.P
Custom Geo Output => [all] [query] [city] [region] [country] [zip] [isp]
.
.P
Example: geo \-a 8\.8\.8\.8 \-o city,zip,isp
.
.IP "\(bu" 4
\-o [options] Returns Specific Geodata
.
.IP "\(bu" 4
\-a [address] For specific ip in \-s
.
.IP "\(bu" 4
\-v Returns Version
.
.IP "\(bu" 4
\-h Returns Help Screen
.
.IP "\(bu" 4
\-u Updates Bash\-Snippets
.
.IP "" 0
.
.SS "Gist"
Description: Manage your gists with git and Github API v3
.
.P
Usage: gist [command] [args]
.
.IP "\(bu" 4
[star|all] List your gists, use \'star\' as argument for your starred gists, \'all\' or \'a\' for both your and starred gists\. Format for each line is: \fIINDEX\fR \fIURL\fR \fIFILE_NUM\fR \fICOMMENT_NUM\fR \fIDESCRIPTION\fR
.
.IP "\(bu" 4
fetch Update the local list of your gists, \'star\' as argument for your starred gists
.
.IP "\(bu" 4
\fIINDEX\fR Show the path of local gist repo and do custom actions
.
.IP "\(bu" 4
new Create a new gist with files or STDIN
.
.IP "\(bu" 4
grep Grep gists by description, filename and content with a given pattern
.
.IP "\(bu" 4
tag Modify/List tags for a gist
.
.IP "\(bu" 4
tags List all tags and pinned tags
.
.IP "\(bu" 4
pin, Pin/Unpin tags, or filter gists with pinned tags
.
.IP "\(bu" 4
lan Filter gists or list gist with coding languages
.
.IP "\(bu" 4
detail Show the detail of a gist
.
.IP "\(bu" 4
edit Edit description for a gist
.
.IP "\(bu" 4
delete Delete gists by given indices
.
.IP "\(bu" 4
push Push changes by git
.
.IP "\(bu" 4
clean Clean local repos of removed gists
.
.IP "\(bu" 4
config Do configuration
.
.IP "\(bu" 4
user Get list of gists with a given Github user
.
.IP "\(bu" 4
github Import selected gist as a new Github repo
.
.IP "\(bu" 4
help Show the helper message
.
.IP "" 0
.
.P
Example:
.
.IP "\(bu" 4
gist (Show your gists)
.
.IP "\(bu" 4
gist tag (Show your gists with tags)
.
.IP "\(bu" 4
gist fetch (update the list of gists from github\.com)
.
.IP "\(bu" 4
gist 3 (show the repo path of your 3rd gist, and do custom actions)
.
.IP "\(bu" 4
gist 3 \-\-no\-action (show the repo path of your 3rd gist, and do not perform actions)
.
.IP "\(bu" 4
gist new \-\-desc bar foo (create a new gist with files and description)
.
.IP "" 0
.
.SS "Lyrics"
Description: Fetch lyrics for a certain song\.
.
.P
Usage: lyrics [flags] or tool [\-a] [arg] [\-s] [arg]
.
.IP "\(bu" 4
\-a Artist of the song to fetch lyrics for
.
.IP "\(bu" 4
\-s Song of the artist to fetch lyrics for
.
.IP "\(bu" 4
\-f Export the lyrics to file rather than outputting to stdout
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
lyrics \-a logic \-s run it
.
.IP "\(bu" 4
lyrics \-a logic \-s run it \-f ~/runItLyrics\.txt
.
.IP "" 0
.
.SS "Meme"
Description: A lightning fast meme generator
.
.P
Usage: tool [flags] or tool [flags] [arguments]
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-f Choose the output filename
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
meme \-f dogeMeme
.
.IP "" 0
.
.SS "Movies"
Description: Provides relevant information about a certain movie\.
.
.P
Usage: movies [flag] or movies [movieToSearch]
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "\(bu" 4
\-d Show detailed information
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
movies Argo
.
.IP "\(bu" 4
movies Inception
.
.IP "" 0
.
.SS "Newton"
Description: Performs numerical calculations all the way up to symbolic math parsing\.
.
.P
Usage: newton [operation] [expression] or newton [flag]
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Valid Symbols: + add \- subtract [ left parenthesis (you must use brackets bash has a bultin for parenthesis) ] right parenthesis (you must use brackets bash has a bultin for parenthesis) * multiply / divide ^ power : between the range of left and right side (only for area under curve) ~ parameter on right side (only for area, tangent line and log)
.
.P
Examples:
.
.IP "\(bu" 4
newton simplify [[2x]+7]*[4x]
.
.IP "\(bu" 4
movies derive x+2x
.
.IP "" 0
.
.SS "Pwned"
Usage: pwned [email]
.
.P
Description: Checks where an email account has been breached
.
.IP "" 4
.
.nf

* \-u Update Bash\-Snippet Tools

* \-h Show the help

* \-v Get the tool version
.
.fi
.
.IP "" 0
.
.P
Examples:
.
.IP "" 4
.
.nf

* pwned emailtocheck@email\.com
.
.fi
.
.IP "" 0
.
.SS "Qrify"
Usage: qrify [stringtoturnintoqrcode]
.
.P
Description: Converts strings or urls into a qr code\.
.
.IP "" 4
.
.nf

* \-u Update Bash\-Snippet Tools

* \-m Enable multiline support

* \-f Save the QRCode to a file

* \-d Decode the QR code from a PNG/GIF/JP(E)G file

* \-h Show the help

* \-v Get the tool version
.
.fi
.
.IP "" 0
.
.P
Examples:
.
.IP "" 4
.
.nf

* qrify this is a test string

* qrify \-m two \en lines

* qrify \-f fileNameHere Insert what you want to encode here

* qrify github\.com # notice no http:// or https:// this will fail

* qrify \-d fileName\.png
.
.fi
.
.IP "" 0
.
.SS "Short"
Description: Unmasks shortended urls\.
.
.P
Usage: short [shortURL] or short [flag]
.
.IP "\(bu" 4
\-e Expand the following URL
.
.IP "\(bu" 4
\-s Shorten the following URL
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Example:
.
.IP "\(bu" 4
Input: short \-e tinyurl\.com/jhkj
.
.IP "\(bu" 4
Output: http://possiblemaliciouswebsiteornot\.com
.
.IP "\(bu" 4
Input: short \-s google\.com
.
.IP "\(bu" 4
Output: http://ki\.tc/23sd1d
.
.IP "" 0
.
.SS "Siteciphers"
Description: Checks the available ciphers for the SSL of an https site\.
.
.P
Usage: siteciphers [flag] or siteciphers [optionalDFlag] [website]
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "\(bu" 4
\-d Set the delay between requests sent to the site (default is 1 sec)
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
siteciphers github\.com
.
.IP "\(bu" 4
siteciphers \-d 0\.75 google\.com
.
.IP "" 0
.
.SS "Stocks"
Description: Finds the latest information on a certain stock\.
.
.P
Usage: stocks [flag] or stocks [company/ticker]
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
stocks AAPL
.
.IP "\(bu" 4
stocks Tesla
.
.IP "" 0
.
.SS "Taste"
Description: A recommendation engine that provides 3 similar items based on some input topic\. Taste also has the ability to provide information on the item of interest\. Supports: shows, books, music, artists, movies, authors, games
.
.P
Usage: taste [flag] [item]
.
.IP "\(bu" 4
\-i Get more information on similar items
.
.IP "\(bu" 4
\-s Get information on the item itself
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
taste \-i Kendrick Lamar
.
.IP "\(bu" 4
taste Catcher in the Ry
.
.IP "\(bu" 4
taste \-s Red Hot Chili Peppers
.
.IP "" 0
.
.SS "Todo"
Description: A simplistic commandline todo list\.
.
.P
Usage: todo [flags] or todo [flags] [arguments]
.
.IP "\(bu" 4
\-c Clear all the current tasks
.
.IP "\(bu" 4
\-r Remove the following task number
.
.IP "\(bu" 4
\-g Get the current tasks
.
.IP "\(bu" 4
\-a Add the following task
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
todo \-a My very first task
.
.IP "\(bu" 4
todo \-r 1
.
.IP "\(bu" 4
todo \-g
.
.IP "\(bu" 4
todo \-c
.
.IP "" 0
.
.SS "Transfer"
Description: Quickly transfer files from the command line\.
.
.P
Usage: transfer [flags] or transfer [flag] [args] or transfer [filePathToUpload]
.
.IP "\(bu" 4
\-d Download a single file
.
.IP "" 4
.
.nf

* First arg: Output file directory
* Second arg: File url id
* Third arg: File name
.
.fi
.
.IP "" 0

.
.IP "\(bu" 4
\-o Onetime file upload
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
transfer ~/fileToTransfer\.txt
.
.IP "\(bu" 4
transfer ~/firstFileToTransfer\.txt ~/secondFileToTransfer\.txt # upload multiple files at once
.
.IP "\(bu" 4
transfer \-d ~/outputDirectory fileID fileName
.
.IP "\(bu" 4
transfer \-o ~/fileToTransfer\.txt
.
.IP "" 0
.
.SS "Weather"
Description: Provides a 3 day forecast on your current location or a specified location\. With no flags Weather will default to your current location\.
.
.P
Usage: weather or weather [flag] or weather [country] or weather [city] [state]
.
.IP "\(bu" 4
weather [optionalLocation] [i][M] get weather in imperial units, optional M means windspeed in m/s
.
.IP "\(bu" 4
weather [optionalLocation] [m][M] lowercase m gets weather in metric units, optional M means windspeed in m/s
.
.IP "\(bu" 4
weather [Moon] grabs the phase of the moon
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
weather
.
.IP "\(bu" 4
weather Paris m
.
.IP "\(bu" 4
weather Tokyo
.
.IP "\(bu" 4
weather Moon
.
.IP "\(bu" 4
weather mM
.
.IP "" 0
.
.SS "Ytview"
Description: Search and play youtube videos right from the terminal\.
.
.P
Usage: ytview [flag] [string] or ytview [videoToSearch]
.
.IP "\(bu" 4
\-s Searches youtube
.
.IP "\(bu" 4
\-c Shows the latest videos of a channel
.
.IP "\(bu" 4
\-u Update Bash\-Snippet Tools
.
.IP "\(bu" 4
\-h Show the help
.
.IP "\(bu" 4
\-v Get the tool version
.
.IP "" 0
.
.P
Examples:
.
.IP "\(bu" 4
ytview \-s Family Guy Chicken Fight
.
.IP "\(bu" 4
ytview \-c Numberphile
.
.IP "" 0
.
.SH "Updating"
With any of the installed tools you can automate the update by running it with the \-u option or passing in update as the arguments
.
.IP "\(bu" 4
stocks update
.
.IP "" 0
.
.P
or
.
.IP "\(bu" 4
stocks \-u
.
.IP "" 0
.
.P
This will clone the repository and install the new versions of scripts that were installed, if you didn\'t install a certain tool this script will not install the new version of that tool\.
.
.SH "Uninstalling"
If you don\'t have the Bash\-Snippets folder anymore clone the repository: git clone https://github\.com/alexanderepstein/Bash\-Snippets`
.
.P
cd into the Bash\-Snippets directory: \fBcd Bash\-Snippets\fR
.
.IP "" 4
.
.nf

To go through a guided uninstall
.
.fi
.
.IP "" 0
.
.IP "\(bu" 4
\&\./uninstall\.sh
.
.IP "" 0
.
.SH "AUTHOR"
Alexander Epstein Github: https://github\.com/alexanderepstein
.
.SH "Contributors"
Jake Meyer Github: https://github\.com/jakewmeyer
.
.P
Linyos Torovoltos Github: https://gitbub\.com/linyostorovovoltos
.
.SH "License"
MIT License
.
.P
Copyright (c) 2017 Alex Epstein
.
.P
Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
.
.P
The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software\.
.
.P
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT\. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE\.


news/news
#!/usr/bin/env bash
# Author: Navan Chauhan https://github.com/navanchauhan

currentVersion="1.20.0"
configuredClient=""
configuredPython=""

## This function determines which http get tool the system has installed and returns an error if there isnt one
getConfiguredClient()
{
  if  command -v curl &>/dev/null; then
    configuredClient="curl"
  elif command -v wget &>/dev/null; then
    configuredClient="wget"
  elif command -v http &>/dev/null; then
    configuredClient="httpie"
  elif command -v fetch &>/dev/null; then
    configuredClient="fetch"
  else
    echo "Error: This tool reqires either curl, wget, httpie or fetch to be installed." >&2
    return 1
  fi
}

## Allows to call the users configured client without if statements everywhere
httpGet()
{
  case "$configuredClient" in
    curl)  curl -A curl -s "$@" ;;
    wget)  wget -qO- "$@" ;;
    httpie) http -b GET "$@" ;;
    fetch) fetch -q "$@" ;;
  esac
}

getConfiguredPython()
{
  if  command -v python2 &>/dev/null; then
    configuredPython="python2"
  elif command -v python &>/dev/null; then
    configuredPython="python"
  else
    echo "Error: This tool requires python 2 to be installed."
    return 1
  fi
}

if [[ $(uname) != "Darwin" ]]; then
  python()
  {
    case "$configuredPython" in
      python2) python2 "$@" ;;
      python)  python "$@" ;;
    esac
  }
fi

checkInternet()
{
  httpGet github.com > /dev/null 2>&1 || { echo "Error: no active internet connection" >&2; return 1; } # query github with a get request
}

## This function grabs some news and using python parses it
## JSON response to extrapolate the information for storage
getNews()
{
	tag=$( echo $* | sed s/\ /\%20/g )
  info=$(httpGet  "http://beta.newsapi.org/v2/everything?q=$tag&apiKey=5d02fa1a64b24a25ac426594e986500a&language=en") > /dev/null #grab the JSON response
  export PYTHONIOENCODING=utf8 #necessary for python in some cases
  echo $info | python -c "import sys, json; print json.load(sys.stdin)['status']" > /dev/null 2>&1 || { echo "Not a valid news tag"; exit 1; } #checking if we get any information back from the server if not chances are it isnt a valid tag
  # The rest of the code is just extrapolating the data with python from the JSON response
  title=$(echo $info | python -c "import sys, json; print json.load(sys.stdin)['articles'][0]['title']")
  author=$(echo $info | python -c "import sys, json; print json.load(sys.stdin)['articles'][0]['author']")
  description=$(echo $info | python -c "import sys, json; print json.load(sys.stdin)['articles'][0]['description']")
  unset info # done with the JSON response not needed anymore
}

## This function uses all the variables that are set by getNews and
## prints them out to the user in a human readable format
printNews()
{
  echo
  echo "============================================="
  echo "| Title: $title"
  echo "| Author: $author"
  echo "| Description: $description"
  echo "============================================="
  echo
}


update()
{
  # Author: Alexander Epstein https://github.com/alexanderepstein
  # Update utility version 2.2.0
  # To test the tool enter in the defualt values that are in the examples for each variable
  repositoryName="Bash-Snippets" #Name of repostiory to be updated ex. Sandman-Lite
  githubUserName="alexanderepstein" #username that hosts the repostiory ex. alexanderepstein
  nameOfInstallFile="install.sh" # change this if the installer file has a different name be sure to include file extension if there is one
  latestVersion=$(httpGet https://api.github.com/repos/$githubUserName/$repositoryName/tags | grep -Eo '"name":.*?[^\\]",'| head -1 | grep -Eo "[0-9.]+" ) #always grabs the tag without the v option

  if [[ $currentVersion == "" || $repositoryName == "" || $githubUserName == "" || $nameOfInstallFile == "" ]]; then
    echo "Error: update utility has not been configured correctly." >&2
    exit 1
  elif [[ $latestVersion == "" ]]; then
    echo "Error: no active internet connection" >&2
    exit 1
  else
    if [[ "$latestVersion" != "$currentVersion" ]]; then
      echo "Version $latestVersion available"
      echo -n "Do you wish to update $repositoryName [Y/n]: "
      read -r answer
      if [[ "$answer" == [Yy] ]]; then
        cd ~ || { echo 'Update Failed'; exit 1; }
        if [[ -d  ~/$repositoryName ]]; then rm -r -f $repositoryName || { echo "Permissions Error: try running the update as sudo"; exit 1; } ; fi
        echo -n "Downloading latest version of: $repositoryName."
        git clone -q "https://github.com/$githubUserName/$repositoryName" && touch .BSnippetsHiddenFile || { echo "Failure!"; exit 1; } &
        while [ ! -f .BSnippetsHiddenFile ]; do { echo -n "."; sleep 2; };done
        rm -f .BSnippetsHiddenFile
        echo "Success!"
        cd $repositoryName || { echo 'Update Failed'; exit 1; }
        git checkout "v$latestVersion" 2> /dev/null || git checkout "$latestVersion" 2> /dev/null || echo "Couldn't git checkout to stable release, updating to latest commit."
        chmod a+x install.sh #this might be necessary in your case but wasnt in mine.
        ./$nameOfInstallFile "update" || exit 1
        cd ..
        rm -r -f $repositoryName || { echo "Permissions Error: update succesfull but cannot delete temp files located at ~/$repositoryName delete this directory with sudo"; exit 1; }
      else
        exit 1
      fi
    else
      echo "$repositoryName is already the latest version"
    fi
  fi
}

usage()
{
  cat <<EOF
News
Description: Brings you the latest headline from a tag
Usage: news [flag] or news [tag]
  -u  Update Bash-Snippet Tools
  -h  Show the help
  -v  Get the tool version
Examples:
  news Google
  news Donald Trump
EOF
}

if [[ $(uname) != "Darwin" ]]; then getConfiguredPython || exit 1; fi
getConfiguredClient || exit 1


while getopts "uvh" opt; do
  case "$opt" in
    \?) echo "Invalid option: -$OPTARG" >&2
        exit 1
        ;;
    h)  usage
        exit 0
        ;;

bash-snippets.sh
#!/usr/bin/env bash
# Author: Navan Chauhan and Alexander Epstein
declare -a tools=(cheat cloudup crypt cryptocurrency currency geo lyrics meme movies newton pwned qrify short siteciphers stocks taste todo transfer weather ytview)
declare -a validTools=()
currentVersion="1.21.0"
configuredClient=""

## This function determines which http get tool the system has installed and returns an error if there isnt one
getConfiguredClient()
{
  if  command -v curl &>/dev/null; then
    configuredClient="curl"
  elif command -v wget &>/dev/null; then
    configuredClient="wget"
  elif command -v http &>/dev/null; then
    configuredClient="httpie"
  elif command -v fetch &>/dev/null; then
    configuredClient="fetch"
  else
    echo "Error: This tool reqires either curl, wget, httpie or fetch to be installed." >&2
    return 1
  fi
}

httpGet()
{
  case "$configuredClient" in
    curl)  curl -A curl -s "$@" ;;
    wget)  wget -qO- "$@" ;;
    httpie) http -b GET "$@" ;;
    fetch) fetch -q "$@" ;;
  esac
}

grablatestversion()
{
  repositoryName="Bash-Snippets"
  githubUserName="alexanderepstein"
  latestVersion=$(httpGet https://api.github.com/repos/$githubUserName/$repositoryName/tags | grep -Eo '"name":.*?[^\\]",'| head -1 | grep -Eo "[0-9.]+" ) #always grabs the tag without the v option
}

checkInternet()
{
  httpGet github.com > /dev/null 2>&1 || { echo "Error: no active internet connection" >&2; return 1; } # query github with a get request
}

header()
{
  title="Bash Snippets"
  installver="Installed Version: $currentVersion"
  latestver="Latest Version: $latestVersion"
  printf "\t\t\t\t     %s\n" "$title"
  printf "\t\t%s\t\t%s\n" "$latestver" "$installver"

}


toolMenu()
{
  while true; do
    clear
    header
    count=1
    for command in "${validTools[@]}"; do
      if [[ $count -gt 9 ]];then c=$count
      else c="0$count"; fi
      spaces=$((40 - $(echo $command | wc -c)))
      echo -n -e "\t\t\t$c."
      for (( i = 0 ; i < $spaces; i++)) ;do
        printf " "
      done
      echo "$command"
      count=$(( $count + 1 ))
    done
    echo -e -n "\t\tChoose a tool or just press enter to go back: "
    read choice
    if [[ $choice == "" ]];then break; fi
    echo -e -n "\t\tEnter any arguments you want to use with the tool: "
    read args
    clear
    ${validTools[$(($choice-1))]} $args
    exit 0
  done
}

menu()
{
  while true; do
    echo -e "\t\t\t01.\t\t\t\tTools"
    echo -e "\t\t\t02.\t\t\t\tInstallation Check"
    echo -e "\t\t\t03.\t\t\t\tView man page"
    echo -e "\t\t\t04.\t\t\t\tUpdate"
    echo -e "\t\t\t05.\t\t\t\tDonate"
    echo -e "\t\t\t06.\t\t\t\tQuit"
    echo -e -n "\n\t\t\tChoose an option: "
    read choice
    if [[ $choice -gt 6 || $choice -lt 1 ]];then echo "Error invalid option!"; exit 1; fi
    clear
    header
    if   [[ $choice -eq 1 ]];then toolMenu
    elif [[ $choice -eq 2 ]];then installationcheck
    elif [[ $choice -eq 3 ]];then man bash-snippets
    elif [[ $choice -eq 4 ]];then ${validTools[1]} -u
    elif [[ $choice -eq 5 ]];then
      clear
      echo -e "\t\tThanks for thinking of donating, that's pretty cool of you"
      echo -e "\n\t\tCryptocurrency Donation Addresses"
      echo -e "\t\t\tBTC: 38Q5VbH63MtouxHu8BuPNLzfY5B5RNVMDn"
      echo -e "\t\t\tETH: 0xf7c60C06D298FF954917eA45206426f79d40Ac9D"
      echo -e "\t\t\tLTC: LWZ3T19YUk66dgkczN7dRhiXDMqSYrXUV4\n"
      echo -e "\t\tNormal Methods"
      echo -e "\t\t\tVenmo: AlexanderEpstein"
      echo -e "\t\t\tSquare Cash: AlexEpstein\n\n\n\n"
      exit 0
    elif [[ $choice -eq 6 ]]; then exit 0
    fi
  done
}

installationcheck()
{
  validTools=()
  for tool in "${tools[@]}"; do
    if [ -e "/usr/local/bin/$tool" ]; then
      state="present";
      start=51;
      validTools+=($tool);
    else state="absent"; start=52; fi
      if $1; then
        echo -n -e "\t\t$tool"
        spaces=$(($start - $( echo $tool | wc -c)))
        for (( i = 0 ; i < $spaces; i++)) ;do
          printf " "
        done
        echo $state
      fi
  done
  if $1; then
    echo -n "Press enter to go back to the menu"
    read
    clear
    header
  fi
}

clear
checkInternet
getConfiguredClient
grablatestversion
header
installationcheck false
echo
menu
